sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/QueuingBeamFnDataGrpcClient.java

```diff
@@ -0,0 +1,171 @@
+    newConsumer.idc = idc;
+    this.idcs.add(idc);
+    return idc;
+  }
+
+  private boolean AllDone() {
+    boolean allDone = true;
+    for (InboundDataClient idc : idcs) {
+      allDone &= idc.isDone();
+    }
+    return allDone;
+  }
+
+  /**
+   * Drains the internal queue of this class, by waiting for all WindowValues to be passed to thier
+   * consumers. The thread which wishes to process() the elements should call this method, as this
+   * will cause the consumers to invoke element processing. All receive() and send() calls must be
+   * made prior to calling drainAndBlock, in order to properly terminate.
+   */
+  public void drainAndBlock() throws Exception {
+    // Note: We just throw the exception here
+    // TODO review the error handling here
+    while (true) {
+      ConsumerAndData tuple = null;
+      tuple = queue.poll(50, TimeUnit.MILLISECONDS);
+      // TODO should we implement a timeout logic here? What happens if the
+      // putting thread throws an exception? Can we assume the InboundDataClient will be marked
+      // done?
+      if (tuple == null) {
+        continue;
+      } else {
+        // Forward to the consumers who cares about this data.
+        tuple.consumer.accept(tuple.data);
+      }
+
+      // TODO is there a possible race here? Leading to data loss?
+      // Can this be set to done, but more elements come in after?
+      if (AllDone()) {
+        break;
+      }
+    }
+  }
+
+  /**
+   * Creates a {@link CloseableFnDataReceiver} using the provided instruction id and target.
+   *
+   * <p>The provided coder is used to encode elements on the outbound stream.
+   *
+   * <p>Closing the returned receiver signals the end of the stream.
+   *
+   * <p>The returned closeable receiver is not thread safe.
+   */
+  @Override
+  public <T> CloseableFnDataReceiver<WindowedValue<T>> send(
+      Endpoints.ApiServiceDescriptor apiServiceDescriptor,
+      LogicalEndpoint outputLocation,
+      Coder<WindowedValue<T>> coder) {
+    LOG.debug(
+        "Creating output consumer for instruction {} and target {}",
+        outputLocation.getInstructionId(),
+        outputLocation.getTarget());
+    return this.mainClient.send(apiServiceDescriptor, outputLocation, coder);
+  }
+
+  public class QueueingFnDataReceiver<T> implements FnDataReceiver<WindowedValue<T>> {
+    private final FnDataReceiver<WindowedValue<T>> consumer;
+    public InboundDataClient idc;
+
+    public QueueingFnDataReceiver(FnDataReceiver<WindowedValue<T>> consumer) {
+      this.consumer = consumer;
+    }
+
+    @Override
+    public void accept(WindowedValue<T> value) throws Exception {
+      // Note: We just throw the exception here
+      // TODO please review this error handling.
+      ConsumerAndData offering = new ConsumerAndData(this.consumer, value);
+      while (!queue.offer(offering, 50, TimeUnit.MILLISECONDS)) {
+        if (idc.isDone()) {
+          // Discard the element.
```

Comment:
This can only get marked as done in two places:
* either the consumer fails to process the bundle (which is ok to drop the element)
* the caller has finished, (can't happen since the thread is here and not processing the next data packet)

Reply:
ack